Extract API sessions & HTTP requests as lower-level client routines

nolar opened the pull request:
We need to prepare the API client routines for the coming feature of retrying all API requests on server-side/networking errors. It is easier to implement such logic in one place rather than to apply it with decorators and/or composing objects in several places.

As such, the lower-level routines are extracted and all the API session management (authentication, response checking, etc) is shifted down the stack. The lower-level routines only do the generic API requests (GET/POST/PATCH/DELETE/etc) without knowing their purpose. The higher-level routines do additional object manipulation before sending anything to the lower-level API requests.

For convenience and code simplicity, most lower-level methods return a parsed object; all of them implicitly verify the response status.

_Extracted from #788 to reduce the diff size and remove unrelated changes._

nolar committed:
Rename the API listing method for brevity and uniformity

Signed-off-by: Sergey Vasilyev <nolar@nolar.info>

psontag commented on kopf/_cogs/clients/api.py:
```diff
@@ -0,0 +1,213 @@
+        *,
+        payload: Optional[object] = None,
+        headers: Optional[Mapping[str, str]] = None,
+        timeout: Optional[aiohttp.ClientTimeout] = None,
```
Looking through the changes here I noticed that in most places there is no explicit timeout configured for the request being made. Is this purpose? Is the intention to rely on the [default timeout of 5min from aiohttp](https://docs.aiohttp.org/en/stable/client_quickstart.html#timeouts) in those cases? It might be nice to make this configurable for the user?

nolar replied:
Good catch :-) The answer is: yes and no.

Yes — the timeouts are not configured and the default 5m timeout is used everywhere except the watch-streams.

No — the timeouts are to be added in the coming PR (together with or immediately after the PR with retrying on server-side/networking errors). I've just split the huge changeset (which exploded in complexity) into several PRs, one PR per one semantic change.

Something like this (it needs tests & docs, but it works in my draft):

```python
@dataclasses.dataclass
class NetworkingSettings:

    request_timeout: Optional[float] = 5 * 60  # == aiohttp.client.DEFAULT_TIMEOUT
    """
    A timeout for the entire duration of an API request (in seconds).

    The timeout is applied to all short requests (GET/POST/PATCH/DELETE/etc).
    It is not applied to continuous watch-streams (which are also GET);
    for the watch-streams, set one of
    ``settings.watching.client_timeout``/``settings.watching.server_timeout``.
    """

    connect_timeout: Optional[float] = None
    """
    A timeout for the connection & handshake of an API request (in seconds).
    """

    error_backoffs: Union[float, Iterable[float]] = (1, 1, 2, 3, 5, 8, 13, 21)
    """
    How many times and with which delays (seconds) to retry the API errors.

    For more information on the API error retrying, see :doc:`/resilience`.
    """
```

psontag replied:
Awesome 👍

nolar committed:
Extract API sessions & HTTP requests as lower-level client routines

We need to prepare the API client routines for the coming feature of retrying all API requests on server-side/networking errors. It is easier to implement such logic in one place rather than to apply it with decorators and/or composing objects in several places.

As such, the lower-level routines are extracted and all the API session management (authentication, response checking, etc) is shifted down the stack. The lower-level routines only do the generic API requests (GET/POST/PATCH/DELETE/etc) without knowing their purpose. The higher-level routines do additional object manipulation before sending anything to the lower-level API requests.

For convenience and code simplicity, most lower-level methods return a parsed object; all of them implicitly verify the response status.

Signed-off-by: Sergey Vasilyev <nolar@nolar.info>

nolar committed:
Shift the API mocks down to the HTTP client level

Signed-off-by: Sergey Vasilyev <nolar@nolar.info>
